crates/polars-lazy/src/scan/file_list_reader.rs

```diff
@@ -1,73 +1,176 @@
+    SpecificPaths { paths: Vec<PathBuf> },
+    // TODO eventually also support things that aren't paths, e.g. Python
+    // file-like objects.
+}
+
+impl Display for MultipleReaderFactories {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            MultipleReaderFactories::GlobPath { path } => write!(f, "{}", path.to_string_lossy()),
+            MultipleReaderFactories::SpecificPaths { paths } => {
+                write!(f, "[")?;
+                for path in paths {
+                    write!(f, "{}, ", path.to_string_lossy())?;
+                }
+                write!(f, "]")
+            },
+        }
+    }
+}
+
+impl MultipleReaderFactories {
+    /// Create a new ReaderSources with a specific glob path.
+    pub fn new_glob(path: impl AsRef<Path>) -> Self {
+        Self::GlobPath { path: path.as_ref().to_owned() }
+    }
+
+    /// Create a new ReaderSources with a set of specific, non-glob paths.
+    pub fn new_paths(paths: Vec<PathBuf>) -> Self {
+        Self::SpecificPaths { paths }
+    }
+
+    /// Get list of specific reader factories.
+    ///
+    /// Returns [None] if path is not a glob pattern.
+    fn iter_readers(&self, cloud_options: Option<&CloudOptions>) -> PolarsResult<ReaderIterator> {
+        match self {
+            MultipleReaderFactories::GlobPath { path } => {
+                let path_str = path.to_string_lossy();
+                polars_glob(&path_str, cloud_options)
+            },
+            MultipleReaderFactories::SpecificPaths { paths } => {
+                Ok(Box::new(paths.clone().into_iter().map(|path| {
+                    let path_str = path.to_string_lossy();
+                    if is_cloud_url(path_str.as_ref()) {
+                        Ok(ReaderFactory::Remote {
+                            location: path_str.to_string(),
+                        })
+                    } else {
+                        let f = File::open(path)?;
+                        Ok(ReaderFactory::Local {
+                            path,
+                            reader: Box::new(f),
+                        })
+                    }
+                })))
+            },
+        }
+        // if paths.is_empty() {
+        //     let path_str = self.path().to_string_lossy();
+        //     if path_str.contains('*') || path_str.contains('?') || path_str.contains('[') {
+        //         polars_glob(&path_str, self.cloud_options()).map(Some)
+        //     } else {
+        //         Ok(None)
+        //     }
+        // } else {
+        //     polars_ensure!(self.path().to_string_lossy() == "", InvalidOperation: "expected only a single path argument");
+        //     // Lint is incorrect as we need static lifetime.
+        //     #[allow(clippy::unnecessary_to_owned)]
+        //     Ok(Some(Box::new(paths.to_vec().into_iter().map(Ok))))
+        // }
+    }
+}
+/// A LazyFileListReader can be in one of two states: either it has a set of
```

Comment:
While it's true that the two states is not an ideal design, in current pre-PR code there are six potential states, so this is still better!